pkg/controller/etcdproxy/certs.go

```diff
@@ -0,0 +1,262 @@
+package etcdproxy
+
+import (
+	"crypto/x509/pkix"
+	"fmt"
+	"math/rand"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	etcdstoragev1alpha1 "github.com/xmudrii/etcdproxy-controller/pkg/apis/etcd/v1alpha1"
+	"github.com/xmudrii/etcdproxy-controller/pkg/certs"
+)
+
+// annCertificateGenerated is annotating are certificates successfully created and stored in the Secret or ConfigMap.
+const annCertificateGenerated = "etcd.xmudrii.com/certificate-generated"
+
+// setNewEtcdProxyCertificates generates new self-signed server signer and server certificate/key pair.
+// Self-signed server signer cert is stored in APIServer ConfigMap to be used as CA by etcd, while server cert/key pair
+// is stored in Controller Secret to be used by etcdproxy.
+func (c *EtcdProxyController) setNewEtcdProxyCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
```

Comment:
Do I understand correctly: we always create a new CA and new certs, even if the `updateAPIServerServingCAConfigMaps` notices later that all config maps are already filled (i.e. the annotation exists?).

Fixing this would be fine in a follow-up. The problem is that creating CAs and certs takes a lot of entropy from the kernel, and eventually you might starve the whole system creating certs.

Reply:
Yes, that's right. While this is a very bad solution as you mentioned, it helped me to get started.

My plan is to fix this in a follow-up. It's already fixed on `certs-renewal` branch.